suprb/optimizer/rule/mutation.py

```diff
@@ -160,3 +161,52 @@ def min_percentage(self, rule: Rule, random_state: RandomState):
         bounds = rule.match.bounds
         bounds[:, 0] -= random_state.uniform(0, self.sigma[0], size=bounds.shape[0])
         bounds[:, 1] += random_state.uniform(0, self.sigma[1], size=bounds.shape[0])
+
+
+class AdaptiveMutation(RuleMutation):
+    """Start off by using a given operator until the fitness decreased
+       more than self.number_of_worse_iterations times. Then change the mutation operator
+       to Normal
+    """
+
+    def __init__(self, matching_type: MatchingFunction = None,
+                 sigma: Union[float, np.ndarray] = 0.1,
+                 operator: RuleMutation = HalfnormIncrease(),
+                 worse_iteration_tolerance: int = 5,
+                 sigma_normal: Union[float, np.ndarray] = 0.1):
+
+        super().__init__(matching_type=matching_type, sigma=sigma)
+        self.best_elitist_fitness = 0
+        self.number_of_worse_iterations = 0
+        self.worse_iteration_tolerance = worse_iteration_tolerance
+        self.sigma_normal = sigma_normal
+
+        # This is necessary because of sklearn guidelines
+        self.operator = operator
+        self.operator_ = clone(self.operator)
+
+    def __call__(self, rule: Rule, random_state: RandomState) -> Rule:
+        return self.operator_(rule, random_state)
+
+    def adapt(self, elitist_fitness: float):
+        if not isinstance(self.operator_, Normal):
+            if elitist_fitness < self.best_elitist_fitness:
+                self.number_of_worse_iterations += 1
+
+                if self.number_of_worse_iterations > self.worse_iteration_tolerance:
+                    self.operator_ = Normal(matching_type=self.matching_type, sigma=self.sigma_normal)
+            else:
+                self.best_elitist_fitness = elitist_fitness
+                self.operator_ = self.operator
+
+    def unordered_bound(self, rule: Rule, random_state: RandomState):
+        raise TypeError("AdaptiveMutation has no implementation for unordered_bound")
```

Comment:
why not use NotImplementedError if the message conveys exactly that? But I think that we should give more explanation here. It is not implemented because we didn't do it (yet) but because it makes no sense to do it as this is only a wrapper

Reply:
Done